File: js/bruby_bind/convert.js
(function()
{

  let RbValue = BRubyBridge.RbValue;


  BRubyBind.Convert = class
  {

    static convert_call(func, ...args)
    {
      let args_rb = args.map(arg => this.js_to_rb(arg));
      try
      {
        let return_rb = func(...args_rb);
        return this.rb_to_js(return_rb);
      }
      catch (error)
      {
        if (error instanceof BRubyBridge.RbError)
        {
          let rb_value = error.rb_value;
          let error_js;
          if (rb_value.responds_to('js_value')) // error originally thrown in js
            error_js = rb_value.send('js_value');
          else // error originally thrown in rb
            error_js = this.rb_to_js(rb_value);
          throw error_js;
        }
        else
          throw error;
      }
    }


    static rb_to_js(object_rb)
    {
      if (!(object_rb instanceof RbValue))
        return object_rb;

      // auxiliary mappings

      // ...


      // primitive values

      if (object_rb.equal_to(RbValue.nil))
        return null;
      if (object_rb.equal_to(RbValue.false))
        return false;
      if (object_rb.equal_to(RbValue.true))
        return true;
      if (object_rb.is_a('Numeric') || object_rb.responds_to('to_int'))
        return object_rb.to_number();
      if (object_rb.is_a('String') || object_rb.is_a('Symbol') || object_rb.responds_to('to_str'))
        return object_rb.to_string();

      // js type place holders

      if (object_rb.is_a('JavaScript::Undefined'))
        return undefined;

      // TODO JSSymbol
      // TODO JSArray

      // js wrapper
      // JavaScript::Function
      // JavaScript::Exception
      // JavaScript::Object
      // TODO: list all
      if (object_rb.responds_to('js_value'))
        return object_rb.send('js_value');

      // native rb object

      // existing rb wrapper (in js)
      let rb_wrapper = BRubyBind.RubyObject.get_wrapper(object_rb);
      if (rb_wrapper)
        return rb_wrapper;

      let wrapper_class_name;
      // Method, Proc
      if (object_rb.responds_to('to_proc'))
        wrapper_class_name = 'RubyClosure';
      // Class
      else if (object_rb.responds_to('new'))
        wrapper_class_name = 'RubyClass';
      // Exception
      else if (object_rb.responds_to('exception'))
        wrapper_class_name = 'RubyError';
      // rb Object
      else
        wrapper_class_name = 'RubyObject';
      
      let rb_wrapper_class = BRubyBind[wrapper_class_name];
      rb_wrapper = new rb_wrapper_class(object_rb);
      BRubyBind.RubyObject.set_wrapper(object_rb, rb_wrapper);

      return rb_wrapper;
    }


    static js_to_rb(object_js)
    {
      return RbValue.get('BRubyBind::Convert').send('js_to_rb', object_js);
    }


  }
  


})();

File: rb/bruby_bind/convert.rb
module BRubyBind
  module Convert
    include BRubyBridge

    
    def self.rb_to_js(object_rb)
      JSValue["BRubyBind.Convert"].call("rb_to_js", object_rb)
    end


    def self.js_to_rb(object_js)
      return object_js unless object_js.is_a?(JSValue)
      
      case

      # should never get RbWrapper
      #when object_js.is_rb_object?
      #  raise

      # auxiliary mappings

      # ...

      # primitive values
      
      when object_js.undefined?
        return nil
      when object_js.null?
        return nil
      when object_js.false?
        return false
      when object_js.true?
        return true
      when object_js.number?
        is_integer = JSValue["Number"].call("isInteger", object_js).to_boolean
        if is_integer
          return object_js.to_integer
        else
          return object_js.to_float
        end
      when object_js.string?
        return object_js.to_string

      # rb type place holders
      # rb Symbol
      when object_js.instanceof(JSValue["BRubyBind.RubySymbol"])
        return object_js["value"].to_symbol
      # rb Float
      when object_js.instanceof(JSValue["BRubyBind.RubyFloat"])
        return object_js["value"].to_float
      # rb Integer
      when object_js.instanceof(JSValue["BRubyBind.RubyInteger"])
        return object_js["value"].to_integer
      # TODO rb Array
      
      # rb wrapper
      when object_js.has?('rb_value')
        return object_js['rb_value']

      # native js object
      else
        # existing js wrapper (in rb)
        js_wrapper = JavaScript::Object.get_wrapper(object_js)
        return js_wrapper if js_wrapper

        js_wrapper = case
        # TODO js Array
        # js error class
        when object_js["prototype"].instanceof(JSValue["Error"]) || object_js == JSValue["Error"]
          JavaScript::Error.new_class(object_js)
        # js error
        when object_js.instanceof(JSValue["Error"])
          JavaScript::Error.new_error(object_js)
        # js Function
        when object_js.typeof == 'function'
          JavaScript::Function.new(object_js)
        # any other js object
        else
          JavaScript::Object.new(object_js)
        end
        JavaScript::Object.set_wrapper(object_js, js_wrapper)
        return js_wrapper
      end
    end

    
  end
end